crates/runtime/src/dataconnector/mysql.rs

```diff
@@ -186,4 +191,205 @@ impl DataConnector for MySQL {
+        Self { metrics }
+    }
+}
+
+const METRICS: &[MetricSpec] = &[
+    MetricSpec::new("connection_count", MetricType::ObservableGaugeU64)
+        .description("Gauge of active connections to the database server"),
+    MetricSpec::new("connections_in_pool", MetricType::ObservableGaugeU64)
+        .description("Gauge of active connections that are idling in the pool"),
+    MetricSpec::new("active_wait_requests", MetricType::ObservableGaugeU64).description(
+        "Gauge of requests that are waiting for a connection to be returned to the pool",
+    ),
+    MetricSpec::new("create_failed", MetricType::ObservableCounterU64)
+        .description("Counter of connections that failed to be created"),
+    MetricSpec::new(
+        "discarded_superfluous_connection",
+        MetricType::ObservableCounterU64,
+    )
+        .description(
+            "Counter of connections that were closed because there were already enough idle connections in the pool",
+        ),
+    MetricSpec::new("discarded_unestablished_connection", MetricType::ObservableCounterU64)
+        .description(
+            "Counter of connections that were closed because they could not be established",
+        ),
+    MetricSpec::new("dirty_connection_return", MetricType::ObservableCounterU64)
+        .description(
+            "Counter of connections that were returned to the pool but were dirty (ie. open transactions, pending queries, etc)",
+        ),
+    MetricSpec::new("discarded_expired_connection", MetricType::ObservableCounterU64)
+        .description(
+            "Counter of connections that were discarded because they were expired by the pool constraints (i.e. TTL expired)",
+        ),
+    MetricSpec::new("resetting_connection", MetricType::ObservableCounterU64)
+        .description(
+            "Counter of connections that were reset",
+        ),
+    MetricSpec::new("discarded_error_during_cleanup", MetricType::ObservableCounterU64)
+        .description(
+            "Counter of connections that were discarded because they returned an error during cleanup",
+        ),
+    MetricSpec::new("connection_returned_to_pool", MetricType::ObservableCounterU64)
+        .description(
+            "Counter of connections that were returned to the pool",
+        ),
+];
+
+impl MetricsProvider for MySQLMetricsProvider {
+    fn component_type(&self) -> ComponentType {
+        ComponentType::Dataset
+    }
+
+    fn component_name(&self) -> &'static str {
+        "mysql"
+    }
+
+    fn available_metrics(&self) -> &'static [MetricSpec] {
+        METRICS
+    }
+
+    #[allow(clippy::too_many_lines)]
+    fn callback_to_observe_metric(
+        &self,
+        metric: &MetricSpec,
+        attributes: Vec<KeyValue>,
+    ) -> Option<ObserveMetricCallback> {
+        let metrics = Arc::clone(&self.metrics);
+        match metric.name {
+            "connection_count" => Some(ObserveMetricCallback::U64(Box::new(move |instrument| {
```

Comment:
The implementation of the callback. This could potentially be a good candidate for a macro to reduce the boilerplate - but I like the ability to be able to put breakpoints at specific code locations for debugging, and its not that much extra code.